### ISSUE TITLE
Eager loading of datasets

### ISSUE THREAD

**Issue by Thf772** [tensorflow/tensorflow#53613]
**System information**
- TensorFlow version (you are using): 2.6.0, 2.7.0
- Are you willing to contribute it (Yes/No): No (I don't have the required knowledge)

**Describe the feature and the current behavior/state.**

In #53292, I mentioned the inability to load datasets from temporary files and mentioned eager dataset load as a potential solution. This led to the discovery and correction of a bug in `Dataset.cache`.

I still believe that eager dataset load is desirable, since calling `cache` then consuming the dataset to achieve the same result feels at best like a makeshift hack. Hence I'm making this new issue as a more explicit feature request, because I do believe that this feature would be beneficial in some cases.

**Will this change the current api? How?**

The feature I propose is to allow callers of `tf.data.experimental.load` to specify whether they want a lazily-loaded dataset, or an eagerly-loaded dataset (that reads the files only once when `load` is called, and stores the data in memory).

The way I see it, this evolution would add a new optional parameter to `tf.data.experimental.load`. Its default value needs to correspond to lazy loading for backward compatibility.

This would not pose a big issue since 1. it will be backward-compatible and 2. only the experimental API will be modified (and the documentation explicitly says the experimental API may evolve at any time).

**Who will benefit from this feature?**

Anyone who currently needs to cache a loaded dataset to memory, be it because the dataset comes from temporary files or because it is accessed very frequently

**Any Other info.**

As in the previous issue:

The fact that the datasets are lazy-loaded is not documented, it would be nice to add this to the current documentation of `tf.data.experimental.load`.

Also on the subject of documentation, once (if) the eager load option is added, the documentation should warn the user about the potentially high memory usage when eager-loading large datasets.

**Comment by sushreebarsa** [comment]
@Thf772  `tf.data.experimental.load` api is deprecated which will be removed in future versions.Please have a look at this  [link](https://www.tensorflow.org/api_docs/python/tf/data/experimental/load?version=nightly) for more details.Please let us know if it helps?
 Thanks!

**Comment by aaudiber** [comment]
@Thf772 Thanks for the extra context. It is still possible to add optional parameters to non-experimental API, as long as the default behavior doesn't change. That being said, my feeling is that this feature doesn't have general enough applicability to be worth adding to the core API. The functionality can already be achieved with a short helper function:

```python
def load_eager(path):
  """Returns an eagerly-loaded dataset that was saved to the given path. The dataset must fit into memory."""
  dataset = tf.data.Dataset.load(path)
  dataset = dataset.cache()
  for _ in dataset: pass  # Iterate through the dataset to load it into memory
  return dataset
```

In general we try to avoid adding parameters that are only useful in very specific situations, especially if there are other ways to achieve the same functionality.

**Comment by Thf772** [comment]
I understand, but the whole point of this issue was to avoid having to use the `cache` hack for this kind of situation. There may be other scenarii in which it is useful to immediately get an in-memory dataset from a saved file.

If you do not wish to expand the capabilities of the API, maybe it would be wise to document the fact that `load` only performs lazy loading, and to describe the `cache` hack for the cases where eager loading is required.

**Comment by aaudiber** [comment]
I agree it's worth specifying in the documentation that the dataset is loaded in a streaming fashion. @shilpakrish2 can you please update the docs to include this? I think the docstring is also out of date with regards to `element_spec`, now that `element_spec` is no longer required in eager mode.

**Comment by alexandrehsd** [comment]
> @Thf772 Thanks for the extra context. It is still possible to add optional parameters to non-experimental API, as long as the default behavior doesn't change. That being said, my feeling is that this feature doesn't have general enough applicability to be worth adding to the core API. The functionality can already be achieved with a short helper function:
> 
> ```python
> def load_eager(path):
>   """Returns an eagerly-loaded dataset that was saved to the given path. The dataset must fit into memory."""
>   dataset = tf.data.Dataset.load(path)
>   dataset = dataset.cache()
>   for _ in dataset: pass  # Iterate through the dataset to load it into memory
>   return dataset
> ```
> 
> In general we try to avoid adding parameters that are only useful in very specific situations, especially if there are other ways to achieve the same functionality.

Hi, I'm using tensorflow 2.12.0 in google colab and this helper function isn't working for me. I keep receiving the error message:  `NotFoundError: Could not find metadata file [my_file]`. I read the whole thread, including [#53292](https://github.com/tensorflow/tensorflow/issues/53292), and I thought this solution would definitely work. Does someone have some clue of what might be happening? I can give more context if needed.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
